### PR TITLE
Refactor samples

### PR DESCRIPTION
**Pull request by Arcnor**
Hello,

I'm not sure this is a welcome change or not, specially due to the size of it, but I've found the existing samples confusing due to how they mix platform code (GLFW) with IGL code (which is the one I'm normally interested in). Also, each sample does the same thing in certain cases in slightly different ways, so I've abstracted all that.

I've split my changes in two groups of files: `SampleCommon` and `IGLCommon`.
1. `SampleCommon` contains all the code that should be relevant only to the Tiny samples, mostly all the GLFW initialization. This makes it easy to replace with something else as well, like SDL or custom code.
2. `IGLCommon` contains code that in my mind should actually be part of IGL, and not only limited to samples. For now it only contains a way to create a device as well as `getNative[Depth]Drawable()`. I'm not sure why the consumer code should care if the underlying API is OpenGL or Vulkan (or the current platform) to perform those actions. Maybe I'm missing something and in certain cases this code might be different, but at least for the examples there doesn't seem to be any reason for it.

In any case, I'm hoping this will be useful.

### CODE REVIEW

File: samples/desktop/Tiny/SamplesCommon.h [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+
+#include <functional>
+
+namespace igl {
+
+class IDevice;
+
+namespace sample {
+
+enum class Keys {
```
Comment (corporateshark) [inline samples/desktop/Tiny/SamplesCommon.h:18]: I completely understand the motivation to refactor these, but it somewhat goes against the purpose of these demos. The intention behind all the demos in the `samples/` directory is to offer straightforward standalone examples without introducing additional layers or abstractions on top of IGL/GLFW. That's why we haven't factored out any shared code here, and so on. If you're looking for cross-platform wrappers for input and other OS-specific functionalities, there's a set of other examples available at `https://github.com/facebook/igl/tree/main/shell`. Those examples might serve as a better starting point if you're considering introducing abstractions, as suggested in this PR.

**Review by corporateshark — CHANGES_REQUESTED** [review]

File: samples/desktop/Tiny/SamplesCommon.h [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+  Key_LEFT_SUPER,
+  Key_RIGHT_SHIFT,
+  Key_RIGHT_CONTROL,
+  Key_RIGHT_ALT,
+  Key_RIGHT_SUPER,
+  Key_MENU,
+};
+
+enum class KeyMods {
+  None = 0x00,
```
Comment (corporateshark) [inline samples/desktop/Tiny/SamplesCommon.h:144]: There's a macro `None` defined in Linux. This code will not compile on Linux.

File: samples/desktop/Tiny/SamplesCommon.h (hunk 1 above)
Comment (Arcnor) [reply samples/desktop/Tiny/SamplesCommon.h:18]: > I completely understand the motivation to refactor these, but it somewhat goes against the purpose of these demos. The intention behind all the demos in the samples/ directory is to offer straightforward standalone examples without introducing additional layers or abstractions on top of IGL/GLFW. That's why we haven't factored out any shared code here, and so on. 

Well, the purpose of my changes was actually the opposite from what you say, I think. I *don't* want to see any platform code in the examples, the fact that you're using GLFW is just an implementation detail, and it should not be part of them (IMHO of course), it doesn't really help me in the understanding of IGL itself. If I could I'd have no abstraction at all here, but every other graphics library I've used contains some form of "sample harness" in the same vein as the one I've created here for the same reason.

> If you're looking for cross-platform wrappers for input and other OS-specific functionalities, there's a set of other examples available at `https://github.com/facebook/igl/tree/main/shell`. Those examples might serve as a better starting point if you're considering introducing abstractions, as suggested in this PR.

I'll take a look at that other set of examples (also, why are there 2 different sets?), but that makes it sound as I'm introducing the abstractions for the sake of it, but as mentioned, they're the means to an end, not the end in itself 😄.

File: samples/desktop/Tiny/SamplesCommon.h (hunk 2 above)
Comment (Arcnor) [reply samples/desktop/Tiny/SamplesCommon.h:144]: Well, this is obviously trivial to fix, but I'm not sure you want to accept this PR for the reasons stated above? Let me know and I'll just rename this, or please close the PR if you are not expecting to merge.

Thank you for taking the time, I appreciate it.

File: samples/desktop/Tiny/SamplesCommon.h (hunk 1 above)
Comment (corporateshark) [reply samples/desktop/Tiny/SamplesCommon.h:18]: > every other graphics library I've used contains some form of "sample harness" in the same vein as the one I've created here for the same reason.

Then this code here is definitely what you are looking for: https://github.com/facebook/igl/tree/main/shell. If you find issues in there, feel free to create a PR.

> why are there 2 different sets?

As mentioned above, to have both approaches demonstrated - with and without "harness".

**Pull request closed without merging**